api/command.go

```diff
@@ -17,630 +16,386 @@ import (
-		implemented := false
-		for _, cmd := range cmds {
-			bounds := len(cmd)
-			if len(command.Command) < bounds {
-				continue
-			}
-			if command.Command[:bounds] == cmd {
-				implemented = true
-			}
-		}
-		if !implemented {
-			c.Err = commandNotImplementedErr
-			return false
-		}
-	}
+	sr := r.PathPrefix("/commands").Subrouter()
 
-	for _, v := range commands {
+	sr.Handle("/execute", ApiUserRequired(executeCommand)).Methods("POST")
+	sr.Handle("/list", ApiUserRequired(listCommands)).Methods("GET")
 
-		if v(c, command) || c.Err != nil {
-			return true
-		}
-	}
+	sr.Handle("/create", ApiUserRequired(createCommand)).Methods("POST")
+	sr.Handle("/list_team_commands", ApiUserRequired(listTeamCommands)).Methods("GET")
+	sr.Handle("/regen_token", ApiUserRequired(regenCommandToken)).Methods("POST")
+	sr.Handle("/delete", ApiUserRequired(deleteCommand)).Methods("POST")
 
-	return false
+	sr.Handle("/test", ApiAppHandler(testCommand)).Methods("POST")
+	sr.Handle("/test", ApiAppHandler(testCommand)).Methods("GET")
 }
 
-func logoutCommand(c *Context, command *model.Command) bool {
-
-	cmd := cmds["logoutCommand"]
-
-	if strings.Index(command.Command, cmd) == 0 {
-		command.AddSuggestion(&model.SuggestCommand{Suggestion: cmd, Description: c.T("api.command.logout_command.description")})
-
-		if !command.Suggest {
-			command.GotoLocation = "/logout"
-			command.Response = model.RESP_EXECUTED
-			return true
+func listCommands(c *Context, w http.ResponseWriter, r *http.Request) {
+	commands := make([]*model.Command, 0, 32)
+	seen := make(map[string]bool)
+	for _, value := range commandProviders {
+		cpy := *value.GetCommand(c)
+		if cpy.AutoComplete && !seen[cpy.Id] {
+			cpy.Sanatize()
+			seen[cpy.Trigger] = true
+			commands = append(commands, &cpy)
 		}
-
-	} else if strings.Index(cmd, command.Command) == 0 {
-		command.AddSuggestion(&model.SuggestCommand{Suggestion: cmd, Description: c.T("api.command.logout_command.description")})
 	}
 
-	return false
-}
-
-func echoCommand(c *Context, command *model.Command) bool {
-	cmd := cmds["echoCommand"]
-	maxThreads := 100
-
-	if !command.Suggest && strings.Index(command.Command, cmd) == 0 {
-		parameters := strings.SplitN(command.Command, " ", 2)
-		if len(parameters) != 2 || len(parameters[1]) == 0 {
-			return false
-		}
-		message := strings.Trim(parameters[1], " ")
-		delay := 0
-		if endMsg := strings.LastIndex(message, "\""); string(message[0]) == "\"" && endMsg > 1 {
-			if checkDelay, err := strconv.Atoi(strings.Trim(message[endMsg:], " \"")); err == nil {
-				delay = checkDelay
-			}
-			message = message[1:endMsg]
-		} else if strings.Index(message, " ") > -1 {
-			delayIdx := strings.LastIndex(message, " ")
-			delayStr := strings.Trim(message[delayIdx:], " ")
-
-			if checkDelay, err := strconv.Atoi(delayStr); err == nil {
-				delay = checkDelay
-				message = message[:delayIdx]
+	if result := <-Srv.Store.Command().GetByTeam(c.Session.TeamId); result.Err != nil {
```

Comment:
Would be a bit quicker to fire of this DB query before the previous loop, but would only be a marginal improvement so 0/5